libs/vkd3d/memory.c

```diff
@@ -186,87 +185,87 @@ static HRESULT vkd3d_try_allocate_device_memory(struct d3d12_device *device,
     bool budget_sensitive;
     VkResult vr;
 
-    /* buffer_mask / sampled_mask etc will generally take care of this,
-     * but for certain fallback scenarios where we select other memory
-     * types, we need to mask here as well. */
-    type_mask &= device->memory_info.global_mask;
-
     allocate_info.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
     allocate_info.pNext = pNext;
     allocate_info.allocationSize = size;
+    allocate_info.memoryTypeIndex = UINT32_MAX;
 
     while (type_mask)
     {
         uint32_t type_index = vkd3d_bitmask_iter32(&type_mask);
-
-        if ((memory_props->memoryTypes[type_index].propertyFlags & type_flags) != type_flags)
-            continue;
-
-        allocate_info.memoryTypeIndex = type_index;
-
-        budget_sensitive = !!(device->memory_info.budget_sensitive_mask & (1u << type_index));
-        if (budget_sensitive)
+        if ((memory_props->memoryTypes[type_index].propertyFlags & type_flags) == type_flags)
         {
-            type_budget = &memory_info->type_budget[type_index];
-            type_current = &memory_info->type_current[type_index];
-            pthread_mutex_lock(&memory_info->budget_lock);
-            if (*type_current + size > *type_budget)
-            {
-                if (vkd3d_config_flags & VKD3D_CONFIG_FLAG_LOG_MEMORY_BUDGET)
-                {
-                    INFO("Attempting to allocate from memory type %u, but exceeding fixed budget: %"PRIu64" + %"PRIu64" > %"PRIu64".\n",
-                            type_index, *type_current, size, *type_budget);
-                }
-                pthread_mutex_unlock(&memory_info->budget_lock);
-
-                /* If we're out of DEVICE budget, don't try other types. */
-                if (type_flags & optional_flags)
-                    return E_OUTOFMEMORY;
-                else
-                    continue;
-            }
+            allocate_info.memoryTypeIndex = type_index;
+            break;
         }
+    }
 
-        vr = VK_CALL(vkAllocateMemory(device->vk_device, &allocate_info, NULL, &allocation->vk_memory));
+    if (allocate_info.memoryTypeIndex == UINT32_MAX)
+    {
+        FIXME("Found no suitable memory type for requested type_flags #%x.\n", type_flags);
+        return E_OUTOFMEMORY;
+    }
 
-        if (budget_sensitive)
+    /* Once we have found a suitable memory type, only attempt to allocate that memory type.
+     * This avoids some problems by design:
+     * - If we want to allocate DEVICE_LOCAL memory, we don't try to fallback to PCI-e BAR memory by mistake.
+     * - If we want to allocate system memory, we don't try to fallback to PCI-e BAR memory by mistake.
+     * - There is no reasonable scenario where we can expect one memory type to fail, and another memory type
+     *   with more memory property bits set to pass. This makes use of the rule where memory types which are a super-set
```

Comment:
Didn't know that Vulkan spec actually had an ordering guarantee on this. Nice trick!
(Maybe it could cite the section or URL, but the meaning is clear either way)